File: src/inform-1.0.0/src/utilities/coalesce.c
// Copyright 2016-2017 ELIFE. All rights reserved.
// Use of this source code is governed by a MIT
// license that can be found in the LICENSE file.
#include <inform/utilities/coalesce.h>
#include <string.h>

static int compare_ints(void const *a, void const *b)
{
    int x = *(int const*)a;
    int y = *(int const*)b;
    if (x < y) return -1;
    if (y < x) return  1;
    return 0;
}

int inform_coalesce(int const *series, size_t n, int *coal, inform_error *err)
{
    if (series == NULL)
        INFORM_ERROR_RETURN(err, INFORM_ETIMESERIES, 0);
    else if (n == 0)
        INFORM_ERROR_RETURN(err, INFORM_ESHORTSERIES, 0);
    else if (coal == NULL)
        INFORM_ERROR_RETURN(err, INFORM_ETIMESERIES, 0);

    int *tmp = malloc(n * sizeof(int));
    if (tmp == NULL)
    {
        INFORM_ERROR_RETURN(err, INFORM_ENOMEM, 0);
    }
    memcpy(tmp, series, n * sizeof(int));
    qsort(tmp, n, sizeof(int), compare_ints);
    int b = 1;
    for (size_t i = 1; i < n; ++i)
    {
        if (tmp[i] != tmp[i-1]) ++b;
    }

    int *map = malloc(b * sizeof(int));
    if (map == NULL)
    {
        free(tmp);
        INFORM_ERROR_RETURN(err, INFORM_ENOMEM, 0);
    }
    map[0] = tmp[0];
    for (size_t i = 1, j = 1; i < n; ++i)
    {
        if (tmp[i] != tmp[i-1]) map[j++] = tmp[i];
    }
    free(tmp);

    for (size_t i = 0; i < n; ++i)
    {
        int *x = bsearch(series + i, map, b, sizeof(int), compare_ints);
        if (x == NULL)
        {
            free(map);
            INFORM_ERROR_RETURN(err, INFORM_EBIN, 0);
        }
        coal[i] = (int) (x - map);
    }

    free(map);
    return b;
}

File: R/coalesce.R
################################################################################
# Copyright 2017 ELIFE. All rights reserved.
# Use of this source code is governed by a MIT
# license that can be found in the LICENSE file.
################################################################################



################################################################################
#' Coalesce
#'
#' Coalesce a time series into as few contiguous states as possible. The magic
#' of information measures is that the actual values of a time series
#' are irrelevant. For example, \code{[0, 1, 0, 1, 1]} has the same entropy as
#' \code{[2, 9, 2, 9, 9} (possibly up to a rescaling). This give us the freedom
#' to shift around the values of a time series as long as we do not change the
#' relative number of states. This function thus provides a way of "compressing"
#' a time series into as small a base as possible. Many of the measures use the
#' base of the time series to determine how much memory to allocate; the larger
#' the base, the higher the memory usage. It also affects the overall
#' performance as the combinatorics climb exponentially with the base. The two
#' standard usage cases for this function are to reduce the base of a time
#' series or ensure that the states are non-negative. Notice that the encoding
#' that is used ensures that the ordering of the states stays the same.
#'     
#' @param series Vector of the time series to coalesce.
#'
#' @return List giving the coalesced time series and its base.
#'
#' @example inst/examples/ex_coalesce.R
#'
#' @export
#'
#' @useDynLib rinform r_coalesce_
################################################################################
coalesce <- function(series) {
  err <- 0
  b   <- 0

  .check_series(series)
  .check_series_vector(series)

  # Extract length of series
  n <- length(series)

  # Convert to integer vector suitable for C
  xs <- as.integer(series)

  coal <- rep(0, n)
  x    <- .C("r_coalesce_",
             series  = xs,
	     n       = as.integer(n),
	     coal    = as.integer(coal),
	     b       = as.integer(b),
	     err     = as.integer(err))

  if (.check_inform_error(x$err) == 0) {
    coal <- x$coal
    b    <- x$b
  }

  list(series = coal, b = b)
}